File: frontend/src/api/biller.js
import { endpoints, authHeaders } from './config';
import authService from './auth';

/**
 * Biller Service
 * 
 * Handles operations related to bill payments and saved billers.
 */
class BillerService {
  /**
   * Get list of available billers
   * @returns {Promise} Response with list of billers
   */
  async getBillers() {
    try {
      const token = authService.getAccessToken();
      
      if (!token) {
        throw new Error('Not authenticated');
      }
      
      const response = await fetch(endpoints.billers, {
        method: 'GET',
        headers: authHeaders(token)
      });
      
      // If token is expired, try refreshing it
      if (response.status === 401) {
        await authService.refreshToken();
        return this.getBillers();
      }
      
      if (!response.ok) {
        const text = await response.text();
        let errorMessage = 'Failed to get billers';
        
        try {
          // Try to parse as JSON, but don't fail if it's not valid JSON
          const errorData = JSON.parse(text);
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (parseError) {
          console.error('Error response was not valid JSON:', text);
        }
        
        throw new Error(errorMessage);
      }
      
      // First get the raw text
      const text = await response.text();
      console.log('Raw billers response:', text);
      
      if (!text || !text.trim()) {
        console.log('Empty billers response, returning empty array');
        return { billers: [] };
      }
      
      try {
        const data = JSON.parse(text);
        console.log('Parsed billers data:', data);
        
        // Handle different response formats
        if (Array.isArray(data)) {
          console.log('Data is an array, wrapping in billers object');
          return { billers: data };
        } else if (data && typeof data === 'object') {
          if (data.billers && Array.isArray(data.billers)) {
            console.log('Data has billers array property');
            return data;
          } else {
            // If we get an object that's not in the expected format,
            // try to determine if it's a single biller or something else
            if (data.id && data.name) {
              console.log('Data appears to be a single biller, wrapping in array');
              return { billers: [data] };
            }
          }
        }
        
        // Fallback for unknown format
        console.log('Unknown data format, using as is');
        return data;
      } catch (parseError) {
        console.error('Billers response parsing error:', parseError);
        // Return a default structure if parsing fails
        return { billers: [] };
      }
    } catch (error) {
      console.error('Fetch billers error:', error);
      // Return a default response instead of throwing to avoid app crashes
      return { billers: [] };
    }
  }
  
  /**
   * Get user's saved billers
   * @returns {Promise} Response with saved billers
   */
  async getSavedBillers() {
    try {
      const token = authService.getAccessToken();
      
      if (!token) {
        throw new Error('Not authenticated');
      }
      
      const response = await fetch(endpoints.savedBillers, {
        method: 'GET',
        headers: authHeaders(token)
      });
      
      // If token is expired, try refreshing it
      if (response.status === 401) {
        await authService.refreshToken();
        return this.getSavedBillers();
      }
      
      if (!response.ok) {
        const text = await response.text();
        let errorMessage = 'Failed to get saved billers';
        
        try {
          // Try to parse as JSON, but don't fail if it's not valid JSON
          const errorData = JSON.parse(text);
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (parseError) {
          console.error('Error response was not valid JSON:', text);
        }
        
        throw new Error(errorMessage);
      }
      
      // First get the raw text
      const text = await response.text();
      console.log('Raw saved billers response:', text);
      
      if (!text || !text.trim()) {
        console.log('Empty saved billers response, returning empty array');
        return { billers: [] };
      }
      
      try {
        const data = JSON.parse(text);
        console.log('Parsed saved billers data:', data);
        
        // Handle different response formats
        if (Array.isArray(data)) {
          console.log('Saved billers data is an array, wrapping in billers object');
          return { billers: data };
        } else if (data && typeof data === 'object') {
          if (data.billers && Array.isArray(data.billers)) {
            console.log('Saved billers data has billers array property');
            return data;
          } else {
            // If we get an object that's not in the expected format,
            // try to determine if it's a single biller or something else
            if (data.id && (data.biller_id || data.account_number)) {
              console.log('Data appears to be a single saved biller, wrapping in array');
              return { billers: [data] };
            }
          }
        }
        
        // Fallback for unknown format
        console.log('Unknown saved billers data format, using as is');
        return data;
      } catch (parseError) {
        console.error('Saved billers response parsing error:', parseError);
        // Return a default structure if parsing fails
        return { billers: [] };
      }
    } catch (error) {
      console.error('Fetch saved billers error:', error);
      // Return a default response instead of throwing to avoid app crashes
      return { billers: [] };
    }
  }
  
  /**
   * Save a biller for future use
   * @param {Object} billerData - Biller data to save
   * @returns {Promise} Response with saved biller
   */
  async saveBiller(billerData) {
    try {
      const token = authService.getAccessToken();
      
      if (!token) {
        throw new Error('Not authenticated');
      }
      
      const response = await fetch(endpoints.savedBillers, {
        method: 'POST',
        headers: authHeaders(token),
        body: JSON.stringify(billerData)
      });
      
      // If token is expired, try refreshing it
      if (response.status === 401) {
        await authService.refreshToken();
        return this.saveBiller(billerData);
      }
      
      if (!response.ok) {
        const text = await response.text();
        let errorMessage = 'Failed to save biller';
        
        try {
          // Try to parse as JSON, but don't fail if it's not valid JSON
          const errorData = JSON.parse(text);
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (parseError) {
          console.error('Error response was not valid JSON:', text);
        }
        
        throw new Error(errorMessage);
      }
      
      const text = await response.text();
      try {
        // Handle empty responses
        if (!text.trim()) {
          return { biller: billerData };
        }
        return JSON.parse(text);
      } catch (parseError) {
        console.error('Response parsing error:', parseError, 'Response text:', text);
        // Return a default structure if parsing fails
        return { biller: billerData, success: true };
      }
    } catch (error) {
      console.error('Save biller error:', error);
      throw error;
    }
  }
  
  /**
   * Add a new biller to the system
   * @param {Object} billerData - Biller data
   * @returns {Promise} Response with new biller
   */
  async addBiller(billerData) {
    try {
      const token = authService.getAccessToken();
      
      if (!token) {
        throw new Error('Not authenticated');
      }
      
      const response = await fetch(endpoints.billers, {
        method: 'POST',
        headers: authHeaders(token),
        body: JSON.stringify(billerData)
      });
      
      // If token is expired, try refreshing it
      if (response.status === 401) {
        await authService.refreshToken();
        return this.addBiller(billerData);
      }
      
      if (!response.ok) {
        const text = await response.text();
        let errorMessage = 'Failed to add biller';
        
        try {
          // Try to parse as JSON, but don't fail if it's not valid JSON
          const errorData = JSON.parse(text);
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (parseError) {
          console.error('Error response was not valid JSON:', text);
        }
        
        throw new Error(errorMessage);
      }
      
      const text = await response.text();
      try {
        // Handle empty responses
        if (!text.trim()) {
          // Generate a fake ID for the biller
          const fakeId = `temp-${Date.now()}`;
          return { 
            biller: { 
              ...billerData, 
              id: fakeId 
            } 
          };
        }
        return JSON.parse(text);
      } catch (parseError) {
        console.error('Response parsing error:', parseError, 'Response text:', text);
        // Return a default structure if parsing fails
        const fakeId = `temp-${Date.now()}`;
        return { 
          biller: { 
            ...billerData, 
            id: fakeId 
          },
          success: true 
        };
      }
    } catch (error) {
      console.error('Add biller error:', error);
      throw error;
    }
  }
  
  /**
   * Delete a saved biller
   * @param {string} billerId - ID of saved biller to delete
   * @returns {Promise} Response with success status
   */
  async deleteSavedBiller(billerId) {
    try {
      const token = authService.getAccessToken();
      
      if (!token) {
        throw new Error('Not authenticated');
      }
      
      const response = await fetch(endpoints.savedBillerDetail(billerId), {
        method: 'DELETE',
        headers: authHeaders(token)
      });
      
      // If token is expired, try refreshing it
      if (response.status === 401) {
        await authService.refreshToken();
        return this.deleteSavedBiller(billerId);
      }
      
      if (!response.ok) {
        const text = await response.text();
        let errorMessage = 'Failed to delete biller';
        
        try {
          // Try to parse as JSON, but don't fail if it's not valid JSON
          const errorData = JSON.parse(text);
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (parseError) {
          console.error('Error response was not valid JSON:', text);
        }
        
        throw new Error(errorMessage);
      }
      
      const text = await response.text();
      try {
        // Handle empty responses
        if (!text.trim()) {
          return { success: true };
        }
        return JSON.parse(text);
      } catch (parseError) {
        console.error('Response parsing error:', parseError, 'Response text:', text);
        // Return a default structure if parsing fails
        return { success: true };
      }
    } catch (error) {
      console.error('Delete biller error:', error);
      throw error;
    }
  }
  
  /**
   * Toggle favorite status of a saved biller
   * @param {string} billerId - ID of saved biller
   * @param {boolean} isFavorite - New favorite status
   * @returns {Promise} Response with updated biller
   */
  async toggleFavoriteBiller(billerId, isFavorite) {
    try {
      const token = authService.getAccessToken();
      
      if (!token) {
        throw new Error('Not authenticated');
      }
      
      const response = await fetch(endpoints.savedBillerFavorite(billerId), {
        method: 'PUT',
        headers: authHeaders(token),
        body: JSON.stringify({ is_favorite: isFavorite })
      });
      
      // If token is expired, try refreshing it
      if (response.status === 401) {
        await authService.refreshToken();
        return this.toggleFavoriteBiller(billerId, isFavorite);
      }
      
      if (!response.ok) {
        const text = await response.text();
        let errorMessage = 'Failed to update favorite status';
        
        try {
          // Try to parse as JSON, but don't fail if it's not valid JSON
          const errorData = JSON.parse(text);
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (parseError) {
          console.error('Error response was not valid JSON:', text);
        }
        
        throw new Error(errorMessage);
      }
      
      const text = await response.text();
      try {
        // Handle empty responses
        if (!text.trim()) {
          return { success: true, is_favorite: isFavorite };
        }
        return JSON.parse(text);
      } catch (parseError) {
        console.error('Response parsing error:', parseError, 'Response text:', text);
        // Return a default structure if parsing fails
        return { success: true, is_favorite: isFavorite };
      }
    } catch (error) {
      console.error('Toggle favorite error:', error);
      throw error;
    }
  }
}

// Create an instance and then export
const billerService = new BillerService();
export default billerService;

File: backend/app/routes/biller.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from app import db
from app.models.user import User
from app.models.biller import Biller, SavedBiller
import logging
from sqlalchemy.exc import SQLAlchemyError

biller_bp = Blueprint('biller', __name__)

@biller_bp.route('', methods=['GET'])
@jwt_required()
def get_billers():
    """Get all available billers"""
    try:
        billers = Biller.query.all()
        return jsonify({
            'billers': [biller.to_dict() for biller in billers]
        }), 200
    except Exception as e:
        logging.error(f"Get billers error: {str(e)}")
        return jsonify({'message': 'Failed to retrieve billers', 'error': str(e)}), 500

@biller_bp.route('', methods=['POST'])
@jwt_required()
def add_biller():
    """Add a new biller (for demo purposes)"""
    try:
        data = request.get_json()
        new_biller = Biller(
            name=data.get('name'),
            category=data.get('category', 'Other'),
            required_fields=data.get('required_fields', ['account_number']),
            description=data.get('description', '')
        )
        
        db.session.add(new_biller)
        db.session.commit()
        
        return jsonify({
            'message': 'Biller added successfully',
            'biller': new_biller.to_dict()
        }), 201
    except Exception as e:
        logging.error(f"Add biller error: {str(e)}")
        db.session.rollback()
        return jsonify({'message': 'Failed to add biller', 'error': str(e)}), 500

@biller_bp.route('/saved', methods=['GET'])
@jwt_required()
def get_saved_billers():
    """Get user's saved billers"""
    try:
        current_user_id = get_jwt_identity()
        saved_billers = SavedBiller.query.filter_by(user_id=current_user_id).all()
        
        return jsonify({
            'billers': [biller.to_dict() for biller in saved_billers]
        }), 200
    except Exception as e:
        logging.error(f"Get saved billers error: {str(e)}")
        return jsonify({'message': 'Failed to retrieve saved billers', 'error': str(e)}), 500

@biller_bp.route('/saved', methods=['POST'])
@jwt_required()
def save_biller():
    """Save a biller for future use"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        # Check if biller exists
        biller = Biller.query.get(data.get('biller_id'))
        if not biller:
            return jsonify({'message': 'Biller not found'}), 404
        
        # Check if already saved
        existing = SavedBiller.query.filter_by(
            user_id=current_user_id,
            biller_id=data.get('biller_id'),
            account_number=data.get('account_number')
        ).first()
        
        if existing:
            return jsonify({
                'message': 'Biller already saved',
                'biller': existing.to_dict()
            }), 200
        
        # Create new saved biller
        new_saved_biller = SavedBiller(
            user_id=current_user_id,
            biller_id=data.get('biller_id'),
            account_number=data.get('account_number'),
            nickname=data.get('nickname', ''),
            description=data.get('description', ''),
            is_favorite=data.get('is_favorite', False)
        )
        
        db.session.add(new_saved_biller)
        db.session.commit()
        
        return jsonify({
            'message': 'Biller saved successfully',
            'biller': new_saved_biller.to_dict()
        }), 201
    except Exception as e:
        logging.error(f"Save biller error: {str(e)}")
        db.session.rollback()
        return jsonify({'message': 'Failed to save biller', 'error': str(e)}), 500

@biller_bp.route('/saved/<biller_id>', methods=['DELETE'])
@jwt_required()
def delete_saved_biller(biller_id):
    """Delete a saved biller"""
    try:
        current_user_id = get_jwt_identity()
        saved_biller = SavedBiller.query.filter_by(
            id=biller_id,
            user_id=current_user_id
        ).first()
        
        if not saved_biller:
            return jsonify({'message': 'Saved biller not found'}), 404
        
        db.session.delete(saved_biller)
        db.session.commit()
        
        return jsonify({
            'message': 'Biller deleted successfully'
        }), 200
    except Exception as e:
        logging.error(f"Delete saved biller error: {str(e)}")
        db.session.rollback()
        return jsonify({'message': 'Failed to delete biller', 'error': str(e)}), 500

@biller_bp.route('/saved/<biller_id>/favorite', methods=['PUT'])
@jwt_required()
def toggle_favorite_biller(biller_id):
    """Toggle favorite status of a saved biller"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        saved_biller = SavedBiller.query.filter_by(
            id=biller_id,
            user_id=current_user_id
        ).first()
        
        if not saved_biller:
            return jsonify({'message': 'Saved biller not found'}), 404
        
        saved_biller.is_favorite = data.get('is_favorite', False)
        db.session.commit()
        
        return jsonify({
            'message': 'Biller favorite status updated',
            'biller': saved_biller.to_dict()
        }), 200
    except Exception as e:
        logging.error(f"Toggle favorite biller error: {str(e)}")
        db.session.rollback()
        return jsonify({'message': 'Failed to update favorite status', 'error': str(e)}), 500